### PR TITLE
Optimize FEPointEvaluation/NM::MappingInfo

### PR DESCRIPTION
**Pull request by bergbauer**
This PR optimizes especially the (fast) linear code path in `FEPointEvaluation` by

- implementing mapping compression for jacobians/inverse jacobians for cartesian/affine cells,
- templating the special case paths for linear polynomials/faces reducing overhead,
- avoiding copies when reading in solution values provided by `FEEvaluation`.

This triples the performance for linear elements in 3D on my workstation and slightly more than doubles it on our high memory bandwidth server. Higher polynomial degrees also profit from the optimizations.

I will break out the changes to `ArrayView` to a separate PR (#15972) as they might be controversial.

Fixes a bug in `evaluate_tensor_product_value_and_gradient_linear()` coming from https://github.com/dealii/dealii/commit/0ede535509660ccf0a3ece46177595b1b7176884 via https://github.com/dealii/dealii/pull/15303.

FYI @kronbichler @peterrum

### CODE REVIEW

File: include/deal.II/matrix_free/fe_point_evaluation.h [hunk 1]
```diff
@@ -2037,13 +2038,21 @@ FEPointEvaluation<n_components_, dim, spacedim, Number>::compute_evaluate_fast(
...
               if constexpr (n_components == 1)
-                result =
-                  internal::evaluate_tensor_product_value_and_gradient_linear<
-                    dim,
-                    scalar_value_type,
-                    VectorizedArrayType,
-                    1,
-                    stride_view>(solution_values.data(), unit_point_ptr[qb]);
+                {
+                  if (fe->n_components() == 1)
```
Comment (bergbauer) [inline include/deal.II/matrix_free/fe_point_evaluation.h:2042]: I am not happy about this additional runtime check but it is necessary to fix the failing test. Any ideas how to avoid this? @peterrum @kronbichler

Comment (peterrum) [reply include/deal.II/matrix_free/fe_point_evaluation.h:2042]: Due to renumbering?

Comment (bergbauer) [reply include/deal.II/matrix_free/fe_point_evaluation.h:2042]: Correct, renumbering is the problem.

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_point_evaluation.h:2042]: The option is to fit this into the `linear` template argument that you resolve once in the beginning: We only go into the fast path with linear shape functions in case `fe->n_components() == 1` and must go to another function elsewhere. I am not sure if the performance hit to use the generic `evaluate_tensor_product_value_and_gradient_shapes<...>(...)` is acceptable for the case that we have multiple FE components but only pick one, probably one cannot expect optimal performance in that case anyway because we carry around too many unknowns in the local vectors in that case.

Comment (peterrum) [reply include/deal.II/matrix_free/fe_point_evaluation.h:2042]: Set islinear to false during setup?

File: include/deal.II/base/derivative_form.h [hunk 2]
```diff
@@ -591,6 +591,82 @@ transpose(const DerivativeForm<1, dim, spacedim, Number> &DF)
...
+inline Tensor<1, spacedim, typename ProductType<Number1, Number2>::type>
+apply_diagonal_transformation(
+  const DerivativeForm<1, dim, spacedim, Number1> &grad_F,
+  const Tensor<1, dim, Number2>                   &d_x)
+{
+  Assert(dim == spacedim,
+         ExcMessage("Only dim = spacedim allowed for diagonal transformation"));
+  Tensor<1, spacedim, typename ProductType<Number1, Number2>::type> dest;
+  for (unsigned int i = 0; i < spacedim; ++i)
+    dest[i] = grad_F[i][i] * d_x[i];
```
[the remark this answers is not in the document]
Comment (bergbauer) [reply include/deal.II/base/derivative_form.h:610]: I think this function has to compile for `dim < spacedim` but throw a runtime error. See the assert above.

File: include/deal.II/base/derivative_form.h [hunk 3]
```diff
@@ -591,6 +591,82 @@ transpose(const DerivativeForm<1, dim, spacedim, Number> &DF)
...
+  return dest;
+}
+
+
+/**
+ * Similar to the previous apply_diagonal_transformation(), specialized for the
+ * case `dim == spacedim` where we can return a rank-2 tensor instead of the
+ * more general `DerivativeForm`.
```
[the remark this answers is not in the document]
Comment (bergbauer) [reply include/deal.II/base/derivative_form.h:618]: This comment is copied form the corresponding `apply_transformation()` function above. This function should only be compiled with `dim == spacedim`.

File: include/deal.II/matrix_free/tensor_product_kernels.h [hunk 4]
```diff
@@ -2724,25 +2723,25 @@ namespace internal
         if (n_values > 1)
           {
             const Number3 tmp0_2 =
-              Number3(values_2[0]) + p[0] * (values_2[1] - values_2[0]);
+              Number3(values[4]) + p[0] * (values[5] - values[4]);
             const Number3 tmp1_2 =
-              Number3(values_2[2]) + p[0] * (values_2[3] - values_2[0]);
+              Number3(values[6]) + p[0] * (values[7] - values[6]); // BUG!!!!
```
[the remark this answers is not in the document]
Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h:2728]: Sure, I marked it to have this bugfix documented. Thanks for providing the PR/commit which introduced this. I will remove the comment.

File: include/deal.II/matrix_free/fe_point_evaluation.h (hunk 1 above)
Comment (bergbauer) [reply include/deal.II/matrix_free/fe_point_evaluation.h:2042]: Thanks for the MPI parallel hint @peterrum @kronbichler ! 🧠 🧠

**Comment by peterrum** [comment]
@bergbauer Could you rebase this PR. I will take a look after that.

**Comment by bergbauer** [comment]
Rebased! @peterrum

File: include/deal.II/matrix_free/fe_point_evaluation.h [hunk 5]
```diff
@@ -847,7 +861,36 @@ class FEPointEvaluation
    */
   void
   evaluate(const ArrayView<const ScalarNumber>    &solution_values,
-           const EvaluationFlags::EvaluationFlags &evaluation_flags);
+           const EvaluationFlags::EvaluationFlags &evaluation_flags)
+  {
+    evaluate<1>(StridedArrayView<const ScalarNumber, 1>(solution_values.data(),
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_point_evaluation.h:866]:
```suggestion
    evaluate(StridedArrayView<const ScalarNumber, 1>(solution_values.data(),
```

File: include/deal.II/matrix_free/fe_point_evaluation.h [hunk 6]
```diff
@@ -1171,6 +1260,23 @@ class FEPointEvaluation
...
+   */
+  std::vector<scalar_value_type> solution_renumbered_face;
+
+  /**
+   * Temporary array to store a vectorized version of the `solution_values`
+   * computed during `integrate()` in a format compatible with the tensor
+   * product evaluators. For vector-valued setups, this array uses a
+   * `Tensor<1, n_components, VectorizedArrayType>` format.
+   */
+  AlignedVector<vectorized_value_type> solution_renumbered_vectorized_face;
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_point_evaluation.h:1278]: Why is this needed? Cannot you use `solution_renumbered` and `solution_renumbered_vectorized`?

File: include/deal.II/matrix_free/fe_point_evaluation.h [hunk 7]
```diff
@@ -2095,23 +2265,40 @@ FEPointEvaluation<n_components_, dim, spacedim, Number>::evaluate(
...
+            evaluate_fast<true, true>(solution_values, evaluation_flags);
+          else
+            evaluate_fast<true, false>(solution_values, evaluation_flags);
+        }
+      else
+        {
+          if (use_linear_path)
+            evaluate_fast<false, true>(solution_values, evaluation_flags);
+          else
+            evaluate_fast<false, false>(solution_values, evaluation_flags);
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_point_evaluation.h:2281]: Not nice. But does the job :+1:

File: include/deal.II/non_matching/mapping_info.h [hunk 8]
```diff
@@ -552,7 +625,7 @@ namespace NonMatching
     /**
      * Offset to point to the first unit point of a cell/face.
      */
-    AlignedVector<unsigned int> unit_points_index;
+    std::vector<unsigned int> unit_points_index;
```
Comment (peterrum) [inline include/deal.II/non_matching/mapping_info.h:628]: This is correct. But was there are particular (performance) reason for this change?

File: include/deal.II/non_matching/mapping_info.h [hunk 9]
```diff
@@ -179,6 +182,49 @@ namespace NonMatching
           }
       }
     };
+
+    template <int dim, int spacedim = dim>
+    dealii::internal::MatrixFreeFunctions::GeometryType
+    compute_geometry_type(
```
Comment (peterrum) [inline include/deal.II/non_matching/mapping_info.h:188]: Is this copy/paste from the matrix-free code?

File: include/deal.II/non_matching/mapping_info.h [hunk 10]
```diff
@@ -885,32 +997,97 @@ namespace NonMatching
...
+                dealii::internal::MatrixFreeFunctions::affine;
+
+            // create a comparator to compare inverse Jacobian of current
+            // and last cell
+            FloatingPointComparator<double> comparator(
+              1e4 / cell->diameter() * std::numeric_limits<double>::epsilon() *
+              1024.);
+
+            // we can only compare if current and last cell have at least
+            // one quadrature point and both cells are at least affine
```
Comment (peterrum) [inline include/deal.II/non_matching/mapping_info.h:1046]: Are you planing to implement something more sophisticated?

File: include/deal.II/non_matching/mapping_info.h [hunk 11]
```diff
@@ -1447,7 +1611,7 @@ namespace NonMatching
   MappingInfo<dim, spacedim, Number>::get_unit_point(
     const unsigned int offset) const
   {
-    return &unit_points[offset];
+    return unit_points.data() + offset;
```
Comment (peterrum) [inline include/deal.II/non_matching/mapping_info.h:1614]: I like this change. But I hope the compiler creates the same code for both cases!?

File: include/deal.II/non_matching/mapping_info.h (hunk 8 above)
Comment (bergbauer) [reply include/deal.II/non_matching/mapping_info.h:628]: Actually, mainly because of pretty printing in gdb. For indices `AligendVector` does not make sense but we also use it in MatrixFree MappingInfo that's why I used it here in the first place.

File: include/deal.II/non_matching/mapping_info.h (hunk 10 above)
Comment (bergbauer) [reply include/deal.II/non_matching/mapping_info.h:1046]: Yes I do, but not at the moment. The next step will be to detect local refinements.

File: include/deal.II/non_matching/mapping_info.h (hunk 11 above)
Comment (bergbauer) [reply include/deal.II/non_matching/mapping_info.h:1614]: This change was motivated by the code in `do_reinit()` that sets the pointers unconditionally in release mode. Otherwise you get segfaults if entry 0 does not exist.

File: include/deal.II/non_matching/mapping_info.h (hunk 9 above)
Comment (bergbauer) [reply include/deal.II/non_matching/mapping_info.h:188]: Copied and adapted. We should aim to reuse this piece of code. Let me see if I can make it fit in MatrixFree MappingInfo as well.

File: include/deal.II/non_matching/mapping_info.h (hunk 10 above)
Comment (kronbichler) [reply include/deal.II/non_matching/mapping_info.h:1046]: For the time when we start thinking about extending the code here, I suggest we try to unify this with the regular `MatrixFreeFunctions::MappingInfo` because most of the functionality there, in particular https://github.com/dealii/dealii/blob/605a6268901ad87cec3d828e742725723298051b/include/deal.II/matrix_free/mapping_info.templates.h#L1066-L1073 which has an `std::map` to detect also long-range dependencies, e.g. in extruded meshes.

File: include/deal.II/matrix_free/fe_point_evaluation.h [hunk 12]
```diff
@@ -1032,93 +1101,113 @@ class FEPointEvaluation
...
-    const ArrayView<const ScalarNumber>    &solution_values,
-    const EvaluationFlags::EvaluationFlags &evaluation_flags);
+    const StridedArrayView<const ScalarNumber, stride_view> &solution_values,
+    const EvaluationFlags::EvaluationFlags                  &evaluation_flags);
 
   /**
    * Evaluates the actual interpolation on the cell or face for a quadrature
    * batch.
    */
+  template <bool is_face_path, bool is_linear, size_t stride_view>
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_point_evaluation.h:1122]: Can we uniformly use 
```suggestion
  template <bool is_face_path, bool is_linear, std::size_t stride_view>
```

**Review by kronbichler — APPROVED** [review]
Nice, this looks good. This was a great effort that you tracked down a lot of unnecessary operations to reduce the overhead in the evaluation routines, and I think the performance numbers you reached with these changes are impressive :+1:

File: include/deal.II/matrix_free/fe_point_evaluation.h (hunk 6 above)
Comment (peterrum) [reply include/deal.II/matrix_free/fe_point_evaluation.h:1278]: @bergbauer

Comment (bergbauer) [reply include/deal.II/matrix_free/fe_point_evaluation.h]: Sorry, we were at Oktoberfest yesterday. You are right, currently `dofs_per_component >= 2 * dofs_per_component_face` which works even for the linear case. Afterward, we only use pointers in the kernels and don't need the correct size. We could of course play safe and resize to `std::max(size_face, size_cell) `during setup. I have removed the additional vectors.

**Review by peterrum — APPROVED** [review]
Thank you for explanation! Good to go from my side!

**Review by kronbichler — APPROVED** [review]
